pkg/gardenlet/controller/shoot/shoot_control.go

```diff
@@ -407,22 +407,24 @@ func (c *Controller) reconcileShoot(logger *logrus.Entry, shoot *gardencorev1bet
 		reconcileAllowed                           = !failedOrIgnored && ((!reconcileInMaintenanceOnly && !confineSpecUpdateRollout(shoot.Spec.Maintenance)) || !isUpToDate || isNowInEffectiveShootMaintenanceTimeWindow)
 	)
 
-	// need retry logic, because the scheduler is acting on it at the same time and cached object might not be up to date
-	updatedShoot, err := kutil.TryUpdateShoot(c.k8sGardenClient.GardenCore(), retry.DefaultBackoff, shoot.ObjectMeta, func(curShoot *gardencorev1beta1.Shoot) (*gardencorev1beta1.Shoot, error) {
-		finalizers := sets.NewString(curShoot.Finalizers...)
-		if finalizers.Has(gardencorev1beta1.GardenerName) {
-			return curShoot, nil
-		}
+	if !controllerutils.HasFinalizer(shoot, gardencorev1beta1.GardenerName) {
+		// need retry logic, because the scheduler is acting on it at the same time and cached object might not be up to date
+		_, err := kutil.TryUpdateShoot(c.k8sGardenClient.GardenCore(), retry.DefaultBackoff, shoot.ObjectMeta, func(curShoot *gardencorev1beta1.Shoot) (*gardencorev1beta1.Shoot, error) {
+			finalizers := sets.NewString(curShoot.Finalizers...)
+			if finalizers.Has(gardencorev1beta1.GardenerName) {
+				return curShoot, nil
+			}
```

Comment:
Can you get rid of this check now in favour of line 410? It looks a little confusing. Alternatively, replace the whole 412-422 with https://github.com/gardener/gardener/blob/9ee208f92d0e3d560f4ee64f688c3a4355b19f20/pkg/controllerutils/finalizers.go#L34 - WDYT?